File: namastack-outbox-examples/namastack-outbox-example-rabbit-java/src/main/java/io/namastack/demo/customer/CustomerService.java
package io.namastack.demo.customer;

import io.namastack.outbox.Outbox;
import jakarta.transaction.Transactional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.UUID;

@Service
public class CustomerService {

    private final CustomerRepository customerRepository;
    private final Outbox outbox;
    private final Logger logger = LoggerFactory.getLogger(CustomerService.class);

    public CustomerService(CustomerRepository customerRepository, Outbox outbox) {
        this.customerRepository = customerRepository;
        this.outbox = outbox;
    }

    @Transactional
    public Customer register(String firstname, String lastname, String email) {
        logger.info("[Service] Register: {} {}", firstname, lastname);

        Customer customer = Customer.register(firstname, lastname, email);
        UUID customerId = customer.getId();

        customerRepository.save(customer);
        logger.info("[Service] Saved to DB: {}", customerId);

        CustomerRegisteredEvent event = new CustomerRegisteredEvent(
            customerId,
            customer.getFirstname(),
            customer.getLastname(),
            customer.getEmail()
        );

        outbox.schedule(event, customerId.toString());
        logger.info("[Service] Scheduled to Outbox: {}", customerId);

        return customer;
    }

    @Transactional
    public void remove(UUID customerId) {
        logger.info("[Service] Remove: {}", customerId);

        customerRepository.deleteById(customerId);

        CustomerRemovedEvent event = new CustomerRemovedEvent(customerId);
        outbox.schedule(event, customerId.toString());

        logger.info("[Service] Scheduled to Outbox: {}", customerId);
    }
}

File: namastack-outbox-examples/namastack-outbox-example-h2/src/main/kotlin/io/namastack/demo/customer/CustomerService.kt
package io.namastack.demo.customer

import io.namastack.outbox.Outbox
import jakarta.transaction.Transactional
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.UUID

@Service
class CustomerService(
    private val customerRepository: CustomerRepository,
    private val outbox: Outbox,
) {
    private val logger = LoggerFactory.getLogger(CustomerService::class.java)

    @Transactional
    fun register(
        firstname: String,
        lastname: String,
        email: String,
    ): Customer {
        logger.info("[Service] Register: {} {}", firstname, lastname)
        val customer = Customer.register(firstname = firstname, lastname = lastname, email = email)
        val customerId = customer.id

        customerRepository.save(customer)
        logger.info("[Service] Saved to DB: {}", customerId)

        outbox.schedule(
            payload =
                CustomerRegisteredEvent(
                    id = customerId,
                    firstname = customer.firstname,
                    lastname = customer.lastname,
                    email = customer.email,
                ),
            key = customerId.toString(),
        )
        logger.info("[Service] Scheduled to Outbox: {}", customerId)

        return customer
    }

    @Transactional
    fun remove(customerId: UUID) {
        logger.info("[Service] Remove: {}", customerId)
        customerRepository.deleteById(customerId)

        outbox.schedule(payload = CustomerRemovedEvent(customerId), key = customerId.toString())
        logger.info("[Service] Scheduled to Outbox: {}", customerId)
    }
}
